relentless/simulate/generic.py

```diff
@@ -0,0 +1,301 @@
+import importlib
+import inspect
+
+import numpy as np
+
+from . import simulate
+
+class SimulationOperationAdapter(simulate.SimulationOperation):
+    """
+    """
+    backends = {}
+
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+        self._op = None
+        self._backend = None
+
+    def __call__(self, sim):
+        if self._op is None or self._backend != sim.backend:
+            backend = self.backends.get(sim.backend)
+            if not backend:
+                raise TypeError('Simulation backend {} not registered.'.format(backend))
+
+            op_name = type(self).__name__
+            try:
+                BackendOp = getattr(backend,op_name)
+            except AttributeError:
+                raise TypeError('{}.{}.{} operation not found.'.format(backend.__module__,backend.__name__,op_name))
+
+            self._op = BackendOp(*self.args,**self.kwargs)
+            self._backend = sim.backend
+
+        return self._op(sim)
+
+    @classmethod
+    def add_backend(cls, backend, module=None):
+        """
+        """
+        # try to deduce module from backend class
+        if module is None:
+            module = backend.__module__
+
+        # setup module (if not already a module)
+        if not inspect.ismodule(module):
+            module = importlib.import_module(module)
+
+        cls.backends[backend] = module
+
+## initializers
+class InitializeFromFile(SimulationOperationAdapter):
+    """Initializes a simulation box and pair potentials from a file.
+
+    Parameters
+    ----------
+    filename : str
+        The file from which to read the system data.
+    options : kwargs
+        Options for file reading.
+    r_buff : float
+        Buffer width (defaults to 0.4).
+
+    """
+    def __init__(self, filename, r_buff=0.4, **options):
+        super().__init__(filename, r_buff, **options)
+
+class InitializeRandomly(SimulationOperationAdapter):
+    """Initializes a randomly generated simulation box and pair potentials.
+
+    Parameters
+    ----------
+    seed : int
+        The seed to randomly initialize the particle locations (defaults to `None`).
+    r_buff : float
+        Buffer width (defaults to 0.4).
+
+    """
+    def __init__(self, seed=None, r_buff=0.4):
+        super().__init__(seed, r_buff)
+
+## integrators
+class MinimizeEnergy(SimulationOperationAdapter):
+    """:py:class:`SimulationOperation` that runs a FIRE energy minimzation until converged.
+
+    Parameters
+    ----------
+    energy_tolerance : float
+        Energy convergence criterion.
+    force_tolerance : float
+        Force convergence criterion.
+    max_iterations : int
+        Maximum number of iterations to run the minimization.
+    dt : float
+        Maximum step size.
```

Comment:
This is likely going to be a `hoomd` specific option, probably should be caught in `**options`.